### PR TITLE
workers: proper TS libs, more spec-compliant APIs

### LINKED ISSUE #3282: Workers can be better
* Support Worker.prototype.terminate(), for terminating workers in main thread;
* Worker extends EventTarget and support `.addEventListener( 'message' )`;
* Support `globalThis.addEventListener( 'message' )` in worker thread;
* Support `self` rather then `window` in worker thread as `globalThis`;
* Support naming:
    ```javascript
    // in main.js
    new Worker( './worker.js', { name:'the name' } );
    ```
    ```javascript
    // in worker.js
    globalThis.name === 'the name';
    ```
* Better support for TypeScript in worker:
    now we must use `window.onmessage`, `window.postMessage` but `globalThis.close` in the worker to prevent error from TypeScript. These APIs need to be unified;
* Support `SharedWorker`;
* ~Support `type: "classic"`~ module is better;
* ~Support `importScripts`~ import and import() are better;

ref: [https://html.spec.whatwg.org/multipage/workers.html](https://html.spec.whatwg.org/multipage/workers.html)

### LINKED ISSUE #3727: worker close API
I'm used to close worker via `globalThis.close()` which I've found from [HTML standard](https://html.spec.whatwg.org/multipage/workers.html#dedicatedworkerglobalscope).
And it works in and before 0.29. When I run the code with 0.30, it throws 'globalThis.close is not a function'. I found my `globalThis.close()` actually crash the worker rather than close in earlier version.  I run `console.log(Object.keys(globalThis))` and found the function I need is `globalThis.workerClose()` in Deno, (there is no documentation in deno.land yet).

I propose that it's better to change the API from `workerClose` to `close` to follow the standard.

### PR DESCRIPTION
**Pull request by bartlomieju**
* split `lib.deno_main.d.ts` into:
  - `lib.deno.shared_globals.d.ts`
  - `lib.deno.window.d.ts`
  - `lib.deno.worker.d.ts`
* remove no longer used libs:
  - `lib.deno_main.d.ts`
  - `lib.deno_worker.d.ts`
* change module loading to use proper TS library for compilation - either "window" or "worker - // NOTE: not happy with current solution, it will be refactored in the near future
* align to Worker API spec:
  - `Worker.terminate()` - currently throws "not implemented" error
  - `self.close()` - renamed from `self.workerClose()`
  - `self.name`

Still a lot to go, but slowly getting there...

Fixes #3727
Fixes #3282

### CODE REVIEW

**Comment by bartlomieju** [comment]
@kitsonk your review would be highly appreciated

File: cli/lib.rs [hunk 1]
```diff
@@ -146,7 +147,13 @@ fn create_worker_and_state(
...
 
 fn types_command() {
-  println!("{}\n{}", crate::js::DENO_NS_LIB, crate::js::DENO_MAIN_LIB);
+  // TODO(bartlomieju): fix types output, after libraries split
+  println!(
+    "{}\n{}\n{}",
+    crate::js::DENO_NS_LIB,
+    crate::js::WINDOW_OR_WORKER_GLOBAL_SCOPE_LIB,
+    crate::js::WINDOW_LIB
+  );
```
[the remark this answers is not in the document]
Comment (bartlomieju) [reply cli/lib.rs]: I'm not sure 😢 CC @kitsonk

File: cli/build.rs [hunk 2]
```diff
@@ -70,13 +70,13 @@ fn main() {
   let bundle_path = o.join("COMPILER_SNAPSHOT.js");
   let snapshot_path = o.join("COMPILER_SNAPSHOT.bin");
   let mut custom_libs: HashMap<String, PathBuf> = HashMap::new();
+  custom_libs
+    .insert("lib.window.d.ts".to_string(), c.join("js/lib.window.d.ts"));
```
Comment (kitsonk) [inline cli/build.rs]: I am not 100% on this name... it implies it is really Deno's cherry picking of the global scope/dom.  I think it should still carry Deno in it somehow... like maybe `lib.deno_window.d.ts`?

File: cli/build.rs [hunk 3]
```diff
@@ -70,13 +70,13 @@ fn main() {
...
+  custom_libs
+    .insert("lib.worker.d.ts".to_string(), c.join("js/lib.worker.d.ts"));
   custom_libs.insert(
-    "lib.deno_main.d.ts".to_string(),
-    c.join("js/lib.deno_main.d.ts"),
-  );
-  custom_libs.insert(
-    "lib.deno_worker.d.ts".to_string(),
-    c.join("js/lib.deno_worker.d.ts"),
+    "lib.window_or_worker_global_scope.d.ts".to_string(),
```
Comment (kitsonk) [inline cli/build.rs]: this feels like a really silly name...  `lib.deno_globals.d.ts`?

File: cli/build.rs [hunk 4]
```diff
@@ -70,13 +70,13 @@ fn main() {
   let bundle_path = o.join("COMPILER_SNAPSHOT.js");
   let snapshot_path = o.join("COMPILER_SNAPSHOT.bin");
   let mut custom_libs: HashMap<String, PathBuf> = HashMap::new();
+  custom_libs
+    .insert("lib.window.d.ts".to_string(), c.join("js/lib.window.d.ts"));
+  custom_libs
+    .insert("lib.worker.d.ts".to_string(), c.join("js/lib.worker.d.ts"));
```
Comment (kitsonk) [inline cli/build.rs]: If this isn't 100% the same as `web_worker` I think we should indicate that it is specifically Deno's version... `lib.deno_worker.d.ts`?

Comment (kitsonk) [reply cli/build.rs]: The other convention that is used in TS libs, would be `js/lib.deno.worker.d.ts`, which is actually maybe better.

File: cli/build.rs [hunk 5]
```diff
@@ -70,13 +70,13 @@ fn main() {
...
-    "lib.deno_main.d.ts".to_string(),
-    c.join("js/lib.deno_main.d.ts"),
-  );
-  custom_libs.insert(
-    "lib.deno_worker.d.ts".to_string(),
-    c.join("js/lib.deno_worker.d.ts"),
+    "lib.window_or_worker_global_scope.d.ts".to_string(),
+    c.join("js/lib.window_or_worker_global_scope.d.ts"),
   );
   custom_libs.insert("lib.deno.d.ts".to_string(), c.join("js/lib.deno.d.ts"));
```
Comment (kitsonk) [inline cli/build.rs]: possibly `lib.deno.ns.d.ts` so it is clear it is the Deno namespace? (I know I am questioning my own code)

File: cli/js/lib.window_or_worker_global_scope.d.ts [hunk 6]
```diff
@@ -6,20 +6,20 @@
 /// <reference lib="deno" />
 /// <reference lib="esnext" />
 
-declare interface Window {
-  window: Window & typeof globalThis;
+// https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope
+
+declare interface WindowOrWorkerGlobalScope {
```
Comment (kitsonk) [inline cli/js/lib.window_or_worker_global_scope.d.ts]: `GlobalScope`?

File: cli/lib.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (kitsonk) [reply cli/lib.rs]: It should be valid, as lib files they are contenatable.  The `/// <reference lib="..." />` may cause some problems though, I haven't validated in VSCode... we might have to strip them.  @axetroy might want to take a look in regards to the VSCode plugin.

File: cli/tests/subdir/test_worker.ts [hunk 7]
```diff
@@ -1,7 +1,11 @@
+if (self.name !== "tsWorker") {
+  throw Error(`Bad worker name: ${self.name}, expected tsWorker`);
```
Comment (kitsonk) [inline cli/tests/subdir/test_worker.ts]:
```suggestion
  throw Error(`Invalid worker name: ${self.name}, expected tsWorker`);
```

File: cli/build.rs (hunk 3 above)
Comment (bartlomieju) [reply cli/build.rs]: Actually it's not my invention: https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope

File: cli/build.rs (hunk 4 above)
Comment (bartlomieju) [reply cli/build.rs]: Not sure about it, we still might want to provide `new Deno.Worker()` which would essentially be Web Worker but with all bells and whistles in `Deno` namespace

File: cli/build.rs (hunk 2 above)
Comment (bartlomieju) [reply cli/build.rs]: Agree, `lib.deno_window.d.ts` SGTM

File: cli/build.rs (hunk 3 above)
Comment (bartlomieju) [reply cli/build.rs]: It's very explicit and descriptive thought, so it would really like to keep it, even if it's silly

File: cli/build.rs (hunk 4 above)
Comment (bartlomieju) [reply cli/build.rs]: Which would essentially be `lib.worker.d.ts` with addition to `Deno` namespace on `self`

File: cli/js/lib.window_or_worker_global_scope.d.ts (hunk 6 above)
Comment (bartlomieju) [reply cli/js/lib.window_or_worker_global_scope.d.ts]: See https://github.com/denoland/deno/pull/3812#discussion_r372130593

File: cli/build.rs (hunk 3 above)
Comment (kitsonk) [reply cli/build.rs]: BAH... You hit me with prior art... Still sometimes we do silly things in standards (e.g. `.getElementById()`)... Just a thought.

File: cli/lib.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (axetroy) [reply cli/lib.rs]: @kitsonk 
I am ok with this.

Splitting a file into multiple files has no effect on Typescript references.

and `/// <reference lib="..." />` This is the official way of references, this should not be a problem.

[the remark this answers is not in the document]
Comment (bartlomieju) [reply cli/lib.rs]: I just added new output of `deno types` to @axetroy's plugin settings and it seems to work fine 👍

[the remark this answers is not in the document]
Comment (axetroy) [reply cli/lib.rs]: @bartlomieju 

I don't have new output of `deno types`. you can try yourself on vscode

```shell
deno types > /path/to/you/deno.d.ts
```

and setting `deno.dtsFilepaths`

```json5
{
  // ...
  "deno.dtsFilepaths": [
    "/path/to/you/deno.d.ts"
  ]
}
```

if it works. then it means ok.

**Pull request merged**